src/main/java/com/limechain/sync/JustificationVerifier.java

```diff
@@ -1,110 +1,148 @@
-//        Ed25519PublicKeyParameters publicKeyParams = new Ed25519PublicKeyParameters(publicKeyBytes, 0);
-//        Ed25519Signer verifier = new Ed25519Signer();
-//        verifier.init(false, publicKeyParams);
-//        verifier.update(data, 0, data.length);
-//
-//        Ed25519PublicKey publicKey =
-//                new Ed25519PublicKey(publicKeyParams);
-//        Extrinsic.ED25519Signature signature = new Extrinsic.ED25519Signature(Hash512.from(signatureHex));
-//
-//        boolean isValid = verifier.verifySignature(signatureBytes);
-//        boolean result = publicKey.verify(data, signature.getValue().getBytes());
-//        if (!result) {
-//            log.log(Level.WARNING, "Invalid signature");
-//        }
-//        return isValid;
-//    }
+    public static boolean verify(Precommit[] precommits, BigInteger round) {
+        SyncState syncState = AppBean.getBean(SyncState.class);
+        Authority[] authorities = syncState.getAuthoritySet();
+        BigInteger authoritiesSetId = syncState.getSetId();
+
+        // Implementation from: https://github.com/smol-dot/smoldot
+        // lib/src/finality/justification/verify.rs
+        if (authorities == null || precommits.length < (authorities.length * 2 / 3) + 1) {
+            log.log(Level.WARNING, "Not enough signatures");
+            return false;
+        }
+
+        Set<Hash256> seenPublicKeys = new HashSet<>();
+        Set<Hash256> authorityKeys =
+                Arrays.stream(authorities).map(Authority::getPublicKey).map(Hash256::new).collect(Collectors.toSet());
+
+        for (Precommit precommit : precommits) {
+            if (!authorityKeys.contains(precommit.getAuthorityPublicKey())) {
+                log.log(Level.WARNING, "Invalid Authority for precommit");
+                return false;
+            }
+
+            if (seenPublicKeys.contains(precommit.getAuthorityPublicKey())) {
+                log.log(Level.WARNING, "Duplicated signature");
+                return false;
+            }
+            seenPublicKeys.add(precommit.getAuthorityPublicKey());
+
+            // TODO (from smoldot): must check signed block ancestry using `votes_ancestries`
+
+            byte[] data = getDataToVerify(precommit, authoritiesSetId, round);
+
+            boolean isValid = verifySignature(
+                    StringUtils.toHex(precommit.getAuthorityPublicKey().getBytes()),
+                    StringUtils.toHex(precommit.getSignature().getBytes()),
+                    StringUtils.toHex(data));
+
+            if (!isValid) {
+                log.log(Level.WARNING, "Failed to verify signature");
+                return false;
+            }
+        }
+        log.log(Level.INFO, "All signatures were verified successfully");
+
+        // From Smoldot implementation:
+        // TODO: must check that votes_ancestries doesn't contain any unused entry
+        // TODO: there's also a "ghost" thing?
```

Comment:
This comment is a bit vague

Reply:
This comment is from the old Fruzhin. I just uncommented the class and added the verifyAsync method.